### PR TITLE
implement lazy loading via Pages' _worker.js directory

### LINKED ISSUE #54: Failed to publish your Function. Got error: Error: Script startup exceeded CPU time limit.
I am using nextjs and prisma data proxy to connect to a database.
Then I deploy to cloudflare pages but failed with `Script startup exceeded CPU time limit`.

I don't know if it is the problem of my project or the work_js generated by next-on-pages.
 
image

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
This change should be ready to go after the following wrangler change is released: https://github.com/cloudflare/workers-sdk/pull/2966

resolves #54

### CODE REVIEW

File: src/buildApplication/generateFunctionsMap.ts [hunk 1]
```diff
@@ -436,3 +433,25 @@ type LooseNode = Node & {
...
+	};
+}
+
+function getFunctionNestingLevel(functionPath: string): number {
+	const relativePath = functionPath.replace(
+		/^.*\/.vercel\/output\/functions\//,
+		''
+	);
+	return relativePath.split('/').length;
+}
```
Comment (GregBrimble) [inline src/buildApplication/generateFunctionsMap.ts:457]: That's some pretty intense path resolution stuff going on there. I think you could avoid this if you were to pre-compile the template, but no biggie.

File: src/buildApplication/generateFunctionsMap.ts [hunk 2]
```diff
@@ -436,3 +433,25 @@ type LooseNode = Node & {
...
+	const accountForNestingPath = `../`.repeat(functionNestingLevel);
+	return chunkKey => {
+		const chunkIdentifier = getChunkIdentifier(chunkKey);
+		const chunkPath = `${accountForNestingPath}__next-on-pages-dist__/chunks/${chunkKey}.js`;
+		return `import ${chunkIdentifier} from '${chunkPath}'`;
+	};
+}
+
+function getFunctionNestingLevel(functionPath: string): number {
+	const relativePath = functionPath.replace(
```
Comment (GregBrimble) [inline src/buildApplication/generateFunctionsMap.ts:452]: https://nodejs.org/api/path.html#pathrelativefrom-to might help here.

File: src/buildApplication/generateFunctionsMap.ts [hunk 3]
```diff
@@ -436,3 +433,25 @@ type LooseNode = Node & {
...
 	value: any;
 };
+
+function getChunkIdentifier(chunkKey: number): string {
+	return `__chunk_${chunkKey}`;
+}
+
+function getChunkImportFn(functionPath: string): (chunkKey: number) => string {
+	const functionNestingLevel = getFunctionNestingLevel(functionPath);
+	const accountForNestingPath = `../`.repeat(functionNestingLevel);
```
Comment (GregBrimble) [inline src/buildApplication/generateFunctionsMap.ts:443]: I'd maybe do something like `path.join(...(new Array(functionNestingLevel).fill('..')))` so you don't need to specify the separator

File: src/buildApplication/generateFunctionsMap.ts [hunk 4]
```diff
@@ -436,3 +433,25 @@ type LooseNode = Node & {
...
+		return `import ${chunkIdentifier} from '${chunkPath}'`;
+	};
+}
+
+function getFunctionNestingLevel(functionPath: string): number {
+	const relativePath = functionPath.replace(
+		/^.*\/.vercel\/output\/functions\//,
+		''
+	);
+	return relativePath.split('/').length;
```
Comment (GregBrimble) [inline src/buildApplication/generateFunctionsMap.ts:456]: Should you use `path.sep` here?

File: src/buildApplication/generateFunctionsMap.ts (hunk 2 above)
Comment (dario-piotrowicz) [reply src/buildApplication/generateFunctionsMap.ts:487]: awesome! thanks! let me give that a try! 🤩👍

File: src/buildApplication/generateFunctionsMap.ts (hunk 3 above)
Comment (dario-piotrowicz) [reply src/buildApplication/generateFunctionsMap.ts:478]: Nice! 🤩👍

File: src/buildApplication/generateFunctionsMap.ts (hunk 2 above)
Comment (dario-piotrowicz) [reply src/buildApplication/generateFunctionsMap.ts]: I've applied the change 🙂  https://github.com/cloudflare/next-on-pages/pull/191/commits/6123deb2b8ca967c93af9e93f2f9b00f33dc1e44

it's a bit more verbose, but I think it's much more clean and robust than before 😄

**Comment by dario-piotrowicz** [comment]
> We might also want to add a peer dep of wrangler@^2.X.Y here or something once my PR gets merged.

I've added a todo comment in the package.json 👍 (that breaks the build so that should help making sure that we don't merge this until I made the change 🙂)

**Comment by corbanvilla** [comment]
@dario-piotrowicz looks like https://github.com/cloudflare/workers-sdk/pull/2966 is merged, anything left in this one before it's ready?

**Comment by dario-piotrowicz** [comment]
@corbanvilla yeah I'll wrap this up as soon as the wrangler release is ready :slightly_smiling_face::+1:

**Pull request merged**